File: src/Blokemon.CardGen/Rendering/CardRenderer.cs
using Blokemon.CardGen.Domain;

namespace Blokemon.CardGen.Rendering;

/// <summary>Prints a card's regions as the shared card markup.</summary>
public static class CardRenderer
{
    private const string _layout = "gym-challenge-750x1050";

    /// <summary>Prints a card.</summary>
    /// <param name="card">The card to print.</param>
    /// <param name="art">The illustrations to embed.</param>
    /// <returns>The card markup.</returns>
    public static string Render(ICard card, Illustrations art)
    {
        ArgumentNullException.ThrowIfNull(card);
        ArgumentNullException.ThrowIfNull(art);

        var regions = string.Concat(card.Regions.Select(region => Region(region, art)));

        return $"""<article class="blokemon-gym-card" data-layout="{_layout}"{Theme(card.ThemeToken)}{Foil(card)} data-canonical-id="{Esc(card.Id.Value)}" aria-label="{Esc(AccessibleName(card))}">{regions}</article>""";
    }

    private static string Theme(string? token) =>
        token is null ? string.Empty : $" data-card-type=\"{token}\"";

    private static string Foil(ICard card) =>
        card
            .Regions.OfType<CardRegion.Colophon>()
            .Any(colophon => colophon.Rarity?.IsHolo() is true)
            ? " data-holo=\"true\""
            : string.Empty;

    private static string AccessibleName(ICard card) =>
        card.Regions.OfType<CardRegion.Vitality>().FirstOrDefault()
            is { Points: { } points } vitality
            ? $"{card.DisplayName}, {points} HP, {vitality.Type} Blokemon"
            : card.DisplayName;

    private static string Region(CardRegion region, Illustrations art) =>
        region.Match(
            PrintedField,
            Nameplate,
            Vitality,
            lineage => Lineage(lineage, art),
            illustration => Illustration(illustration, art),
            Denomination,
            IdentityStrip,
            Mechanics,
            Affinities,
            Colophon
        );

    private static string PrintedField(CardRegion.PrintedField _) =>
        """<div class="printed-inner-field" data-region="inner-field" aria-hidden="true"></div>""";

    private static string Nameplate(CardRegion.Nameplate nameplate) =>
        $"""{Classification(nameplate.Classification)}<h2 class="card-name" style="--blokemon-name-len:{nameplate.Name.Length}">{Esc(nameplate.Name)}</h2>""";

    private static string Classification(string? classification) =>
        classification is null
            ? string.Empty
            : $"""<span class="classification">{Esc(classification)}</span>""";

    private static string Vitality(CardRegion.Vitality vitality) =>
        $"""{HitPoints(vitality.Points)}<span class="main-type-icon" data-energy="{vitality.Type}" aria-label="{vitality.Type} type">{TypeGlyphs.Reference(vitality.Type)}</span>""";

    private static string HitPoints(HitPoints? points) =>
        points is null
            ? string.Empty
            : $"""<div class="hp-cluster"><strong>{points}</strong><small>HP</small></div>""";

    private static string Lineage(CardRegion.Lineage lineage, Illustrations art) =>
        $"""<div class="evolution-panel" data-region="evolution-panel">{art.Image(lineage.Previous.Art, "previous-art")}</div><div class="evolves-strip">Evolves from {Esc(lineage.Previous.Name)}</div>""";

    private static string Illustration(CardRegion.Illustration illustration, Illustrations art) =>
        $"""{Frame(illustration.Placement)}<figure class="art-viewport" data-region="art-viewport" data-placement="{Placement(illustration.Placement)}">{art.Image(illustration.Art, null)}</figure>""";

    private static string Frame(IllustrationPlacement placement) =>
        placement is IllustrationPlacement.Framed
            ? """<div class="art-frame" data-region="art-frame" aria-hidden="true"></div>"""
            : string.Empty;

    private static string Placement(IllustrationPlacement placement) =>
        placement switch
        {
            IllustrationPlacement.Framed => "framed",
            IllustrationPlacement.Field => "field",
            IllustrationPlacement.FullBleed => "full-bleed",
            _ => throw new ArgumentOutOfRangeException(nameof(placement)),
        };

    private static string Denomination(CardRegion.Denomination denomination) =>
        $"""<span class="denomination" data-region="denomination" data-energy="{denomination.Type}" aria-label="{denomination.Type} Energy">{TypeGlyphs.Reference(denomination.Type)}</span>""";

    private static string IdentityStrip(CardRegion.IdentityStrip strip) =>
        $"""<div class="metadata-strip" data-region="metadata-strip"><span>{Esc(strip.Identity)}</span></div>""";

    private static string Mechanics(CardRegion.Mechanics mechanics) =>
        $"""<div class="rules-flow" data-region="rules-flow">{string.Concat(mechanics.Entries.Select(entry => entry.Match(Note, Attack, Note)))}</div><div class="footer-rule" aria-hidden="true"></div>""";

    private static string Affinities(CardRegion.Affinities affinities) =>
        $"""<div class="stat-row" data-region="stat-row"><div class="footer-stat is-weakness"><b>weakness</b>{StatValue(affinities.Weakness)}</div><div class="footer-stat is-resistance"><b>resistance</b>{StatValue(affinities.Resistance)}</div><div class="footer-stat is-retreat"><b>retreat cost</b>{RetreatValue(affinities.Retreat)}</div></div>""";

    private static string Colophon(CardRegion.Colophon colophon) =>
        $"""{Flavour(colophon.Flavour, colophon.Index)}<div class="legal-row" data-region="legal-row">{CollectorMark(colophon.Number)}{RarityImprint(colophon.Rarity)}</div>""";

    private static string Flavour(string? flavour, string? index) =>
        flavour is null
            ? string.Empty
            : $"""<div class="flavour-box" data-region="flavour-box"><p class="flavour">{Esc(flavour)}{Index(index)}</p></div>""";

    private static string Index(string? index) =>
        index is null ? string.Empty : $""" <span class="card-index">{Esc(index)}</span>""";

    private static string CollectorMark(CollectorNumber? number) =>
        number is { } printed
            ? $"""<span class="collector-number">{printed.PrintedLabel()}</span>"""
            : string.Empty;

    private static string RarityImprint(Rarity? rarity) =>
        rarity is { } printed
            ? $"""<span class="rarity-label">{printed.PrintedLabel()}</span><span class="rarity-mark" aria-label="{printed.PrintedLabel()} rarity">{printed.RarityMark()}</span>"""
            : string.Empty;

    private static string Attack(CardEntry.Attack attack) =>
        $"""<section class="rule-entry" data-mechanical-id="{Esc(attack.Id.Value)}" data-entry-kind="Attack"><div class="entry-energy">{string.Concat(attack.EnergyCost.Select(Pip))}</div><p class="entry-copy{NameOnly(attack.EffectText)}"><b class="entry-name">{Esc(attack.Name)}</b>{Esc(attack.EffectText ?? string.Empty)}</p><strong class="entry-damage" aria-label="{attack.Damage} Damage">{attack.Damage}</strong></section>""";

    private static string Note(CardEntry entry)
    {
        var (kind, effect) = entry switch
        {
            CardEntry.Ability ability => ("Ability", ability.EffectText),
            CardEntry.Rule rule => ("Rule", rule.EffectText),
            _ => throw new ArgumentOutOfRangeException(nameof(entry)),
        };

        return $"""<section class="rule-entry is-note" data-mechanical-id="{Esc(entry.Id.Value)}" data-entry-kind="{kind}"><p class="entry-copy"><span class="entry-lead">{kind}:</span><b class="entry-name">{Esc(entry.Name)}</b>{Esc(effect)}</p></section>""";
    }

    private static string NameOnly(string? effectText) =>
        string.IsNullOrWhiteSpace(effectText) ? " is-name-only" : string.Empty;

    private static string Pip(BlokemonType type) =>
        $"""<span class="energy-pip" data-energy="{type}" aria-label="{type} Energy">{TypeGlyphs.Reference(type)}</span>""";

    private static string StatValue(TypeAffinity? affinity) =>
        affinity is null
            ? """<span class="stat-value">&#8212;</span>"""
            : $"""<span class="stat-value"><span class="footer-icon" data-energy="{affinity.Type}" aria-hidden="true">{TypeGlyphs.Reference(affinity.Type)}</span>{Esc(affinity.PrintedValue())}</span>""";

    private static string RetreatValue(RetreatCost retreat) =>
        retreat.Value is 0
            ? """<span class="stat-value">&#8212;</span>"""
            : $"""<span class="stat-value">{string.Concat(Enumerable.Repeat(RetreatPip(), retreat.Value))}</span>""";

    private static string RetreatPip() =>
        $"""<span class="footer-icon" data-energy="Local" aria-label="Local Energy">{TypeGlyphs.Reference(BlokemonType.Local)}</span>""";

    // The card is delivered as XML, so every entity is numeric and every void element is closed.
    private static string Esc(string value) =>
        value
            .Replace("&", "&amp;", StringComparison.Ordinal)
            .Replace("<", "&lt;", StringComparison.Ordinal)
            .Replace(">", "&gt;", StringComparison.Ordinal)
            .Replace("\"", "&quot;", StringComparison.Ordinal)
            .Replace("'", "&#39;", StringComparison.Ordinal);
}

File: src/Blokemon.CardGen/Rendering/CardRenderer.fs
namespace Blokemon.CardGen.Rendering

open System
open Blokemon.CardGen.Domain

/// Prints a card's regions as the shared card markup.
module CardRenderer =

    let private layout = "gym-challenge-750x1050"

    // The card is delivered as XML, so every entity is numeric and every void element is closed.
    let private esc (value: string) =
        value
            .Replace("&", "&amp;", StringComparison.Ordinal)
            .Replace("<", "&lt;", StringComparison.Ordinal)
            .Replace(">", "&gt;", StringComparison.Ordinal)
            .Replace("\"", "&quot;", StringComparison.Ordinal)
            .Replace("'", "&#39;", StringComparison.Ordinal)

    let private theme token =
        match token with
        | None -> ""
        | Some token -> $" data-card-type=\"{token}\""

    let private foil (card: Card) =
        let holo =
            card.Regions
            |> Seq.exists (function
                | CardRegion.Colophon(rarity = Some rarity) -> Rarity.isHolo rarity
                | _ -> false)

        if holo then " data-holo=\"true\"" else ""

    let private vitalityOf (card: Card) =
        card.Regions
        |> Seq.tryPick (function
            | CardRegion.Vitality(points = points; printedType = printedType) ->
                Some(points, printedType)
            | _ -> None)

    let private accessibleName (card: Card) =
        match vitalityOf card with
        | Some(Some points, printedType) ->
            $"{card.DisplayName}, {points} HP, {printedType} Blokemon"
        | _ -> card.DisplayName

    let private classification classification =
        match classification with
        | None -> ""
        | Some label -> $"""<span class="classification">{esc label}</span>"""

    let private nameplate classificationLabel (name: string) =
        $"""{classification classificationLabel}<h2 class="card-name" style="--blokemon-name-len:{name.Length}">{esc name}</h2>"""

    let private hitPoints points =
        match points with
        | None -> ""
        | Some points ->
            $"""<div class="hp-cluster"><strong>{points}</strong><small>HP</small></div>"""

    let private vitality points printedType =
        $"""{hitPoints points}<span class="main-type-icon" data-energy="{printedType}" aria-label="{printedType} type">{TypeGlyphs.reference printedType}</span>"""

    let private lineage (previous: PreviousStage) art =
        let thumbnail =
            IllustrationRendering.image previous.Art IllustrationRole.PreviousStage art

        $"""<div class="evolution-panel" data-region="evolution-panel">{thumbnail}</div><div class="evolves-strip">Evolves from {esc previous.Name}</div>"""

    let private frame placement =
        match placement with
        | IllustrationPlacement.Framed ->
            """<div class="art-frame" data-region="art-frame" aria-hidden="true"></div>"""
        | _ -> ""

    let private placementToken placement =
        match placement with
        | IllustrationPlacement.Framed -> "framed"
        | IllustrationPlacement.Field -> "field"
        | IllustrationPlacement.FullBleed -> "full-bleed"

    let private illustration artwork placement art =
        let drawn = IllustrationRendering.image artwork IllustrationRole.Primary art

        $"""{frame placement}<figure class="art-viewport" data-region="art-viewport" data-placement="{placementToken placement}">{drawn}</figure>"""

    let private denomination printedType =
        $"""<span class="denomination" data-region="denomination" data-energy="{printedType}" aria-label="{printedType} Energy">{TypeGlyphs.reference printedType}</span>"""

    let private identityStrip identity =
        $"""<div class="metadata-strip" data-region="metadata-strip"><span>{esc identity}</span></div>"""

    let private pip (printedType: BlokemonType) =
        $"""<span class="energy-pip" data-energy="{printedType}" aria-label="{printedType} Energy">{TypeGlyphs.reference printedType}</span>"""

    let private nameOnly effectText =
        match effectText with
        | Some text when not (String.IsNullOrWhiteSpace text) -> ""
        | _ -> " is-name-only"

    let private attackEntry (entry: CardEntry) energyCost damage effectText =
        let energy = energyCost |> Seq.map pip |> String.concat ""
        let text = effectText |> Option.defaultValue ""

        $"""<section class="rule-entry" data-mechanical-id="{esc entry.Id.Value}" data-entry-kind="Attack"><div class="entry-energy">{energy}</div><p class="entry-copy{nameOnly effectText}"><b class="entry-name">{esc entry.Name}</b>{esc text}</p><strong class="entry-damage" aria-label="{damage} Damage">{damage}</strong></section>"""

    let private noteEntry (entry: CardEntry) =
        let kind, effect =
            match entry with
            | CardEntry.PokemonPower(effectText = effectText) -> "Blokemon Power", effectText
            | CardEntry.Rule(effectText = effectText) -> "Rule", effectText
            | _ -> raise (ArgumentOutOfRangeException(nameof entry))

        $"""<section class="rule-entry is-note" data-mechanical-id="{esc entry.Id.Value}" data-entry-kind="{kind}"><p class="entry-copy"><span class="entry-lead">{kind}:</span><b class="entry-name">{esc entry.Name}</b>{esc effect}</p></section>"""

    let private entryMarkup entry =
        match entry with
        | CardEntry.Attack(energyCost = energyCost; damage = damage; effectText = effectText) ->
            attackEntry entry energyCost damage effectText
        | CardEntry.PokemonPower _
        | CardEntry.Rule _ -> noteEntry entry

    let private mechanics entries =
        let flow = entries |> Seq.map entryMarkup |> String.concat ""

        $"""<div class="rules-flow" data-region="rules-flow">{flow}</div><div class="footer-rule" aria-hidden="true"></div>"""

    let private statValue affinity =
        match affinity with
        | None -> """<span class="stat-value">&#8212;</span>"""
        | Some(affinity: TypeAffinity) ->
            $"""<span class="stat-value"><span class="footer-icon" data-energy="{affinity.Type}" aria-hidden="true">{TypeGlyphs.reference affinity.Type}</span>{esc (affinity.PrintedValue())}</span>"""

    let private retreatPip () =
        $"""<span class="footer-icon" data-energy="Local" aria-label="Local Energy">{TypeGlyphs.reference BlokemonType.Local}</span>"""

    let private retreatValue (retreat: RetreatCost) =
        if retreat.Value = 0 then
            """<span class="stat-value">&#8212;</span>"""
        else
            let pips = String.replicate retreat.Value (retreatPip ())
            $"""<span class="stat-value">{pips}</span>"""

    let private affinities weakness resistance retreat =
        $"""<div class="stat-row" data-region="stat-row"><div class="footer-stat is-weakness"><b>weakness</b>{statValue weakness}</div><div class="footer-stat is-resistance"><b>resistance</b>{statValue resistance}</div><div class="footer-stat is-retreat"><b>retreat cost</b>{retreatValue retreat}</div></div>"""

    let private index index =
        match index with
        | None -> ""
        | Some index -> $""" <span class="card-index">{esc index}</span>"""

    let private flavour flavourText indexLine =
        match flavourText with
        | None -> ""
        | Some flavourText ->
            $"""<div class="flavour-box" data-region="flavour-box"><p class="flavour">{esc flavourText}{index indexLine}</p></div>"""

    let private collectorMark number =
        match number with
        | None -> ""
        | Some(number: CollectorNumber) ->
            $"""<span class="collector-number">{number.PrintedLabel()}</span>"""

    let private rarityImprint rarity =
        match rarity with
        | None -> ""
        | Some rarity ->
            let printed = Rarity.printedLabel rarity

            $"""<span class="rarity-label">{printed}</span><span class="rarity-mark" aria-label="{printed} rarity">{Rarity.mark rarity}</span>"""

    let private colophon flavourText indexLine rarity number =
        $"""{flavour flavourText indexLine}<div class="legal-row" data-region="legal-row">{collectorMark number}{rarityImprint rarity}</div>"""

    let private printedField () =
        """<div class="printed-inner-field" data-region="inner-field" aria-hidden="true"></div>"""

    let private region art region =
        match region with
        | CardRegion.PrintedField -> printedField ()
        | CardRegion.Nameplate(classification = classification; name = name) ->
            nameplate classification name
        | CardRegion.Vitality(points = points; printedType = printedType) ->
            vitality points printedType
        | CardRegion.Lineage(previous = previous) -> lineage previous art
        | CardRegion.Illustration(art = artwork; placement = placement) ->
            illustration artwork placement art
        | CardRegion.Denomination(printedType = printedType) -> denomination printedType
        | CardRegion.IdentityStrip(identity = identity) -> identityStrip identity
        | CardRegion.Mechanics(entries = entries) -> mechanics entries
        | CardRegion.Affinities(weakness = weakness; resistance = resistance; retreat = retreat) ->
            affinities weakness resistance retreat
        | CardRegion.Colophon(
            flavour = flavourText; index = indexLine; rarity = rarity; number = number) ->
            colophon flavourText indexLine rarity number

    /// Prints a card.
    let render (card: Card) art =
        let regions = card.Regions |> Seq.map (region art) |> String.concat ""

        $"""<article class="blokemon-gym-card" data-layout="{layout}"{theme card.ThemeToken}{foil card} data-canonical-id="{esc card.Id.Value}" aria-label="{esc (accessibleName card)}">{regions}</article>"""
